Blob storage changes

elenasolomon opened the pull request:
- implementation for get/setServiceProperties
- changed the lastModified to be a Date
- small corrections for conditional headers
- corrects the listBlob error message
- adds new tests for SAS
- generalization for authorizeWithRefreshSAS, authorizeWithSAS and authorizeWithSharedKey

I moved the authorizeWithSAS, authorizeWithRefreshSAS and authorizeWithSharedKey in a separate js file because these functions are 90% identical in all the three services.
Please tell me if you agree with this change.

jonasfj commented:
Please make test pass before merging...

jonasfj commented:
For travis you'll have to push a branch to this repo..

jhford commented:
I opened #17 for the tests.  Because it's the exact same changesets as this PR, it looks like github will update this one with the results from that PR.  Nifty!

@elenasolomon which of these changesets are new and as yet unreviewed?

elenasolomon commented:
The changes are in 3bb4664 and 8b9f905 commits and are described in the PR description. Sorry about the long list of commits :(, but the diff is correct and reflects the changes that should be reviewed.

elenasolomon commented:
Meanwhile, I will try to fix the tests

jhford commented:
So version 0.12 is still failing.  I've done some debugging and found that it's related to nvm interacting poorly with libxmljs.  I verified that installing using a binary tarball from nodejs upstream succeeded in installing libxmljs, but that an nvm managed installation of 0.12 does not work.  It is related to the nodejs headers not being correct, or something similar.

This incompatibility is important because Travis uses nvm to install its node environment: https://travis-ci.org/taskcluster/fast-azure-storage/jobs/204127472#L254

Is there any reason why we need to maintain 0.12 support?  Node 0.12 is no longer supported by upstream: https://nodejs.org/en/blog/release/v0.12.18/

I propose a diff like this:
```diff
diff --git a/.travis.yml b/.travis.yml
index 3dbc5dc..0028acc 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -1,7 +1,6 @@
 language: node_js
 sudo: false
 node_js:
-- '0.12'
 - '4'
 - '6'
 addons:
diff --git a/package.json b/package.json
index 21d0f90..5c75485 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name":         "fast-azure-storage",
-  "version":      "0.3.7",
+  "version":      "1.0.0",
   "author":       "Jonas Finnemann Jensen <jopsen@gmail.com>",
   "description":  "Fast client library for azure storage services",
   "license":      "MPL-2.0",
@@ -14,6 +14,9 @@
     "type":       "git",
     "url":        "https://github.com/taskcluster/fast-azure-storage.git"
   },
+  "engines": {
+    "node": ">=4"
+  },
   "dependencies": {
     "debug":                            "^2.2.0",
     "promise":                          "^7.0.4",
```

jhford commented:
Unit tests with 0.12 dropped: https://travis-ci.org/taskcluster/fast-azure-storage/builds/204184596

jhford commented on README.md:
```diff
@@ -175,3 +176,61 @@ See also [reference documentation](https://taskcluster.github.io/fast-azure-stor
  * `Queue#sas(queue, options)`
 
+Azure Blob Storage Client
+--------------------------
```
we're using `#` headers elsewhere, I think we should use them here

elenasolomon replied:
In Markdown there are 2 ways to specify H2 headers.
1) using ## in front of text 
2) using underline style (my case)
I also wanted to be consistent with the other headers from the README document

jhford commented on test/blob_test.js:
```diff
@@ -1,4 +1,4 @@
-suite.only("Azure Blob", function() {
+suite("Azure Blob", function() {
```
I noticed that you're removing `.only` but not moving the sleeps.  Does that mean the sleeps fixed the issue?

elenasolomon replied:
In test 'delete container with if-modified-since conditional header', I was creating a container and after that I wanted to deleted it using the conditional header 'if-modified-since' set to now() date. I was expecting to fail because lastModified < now(), but sometimes lastModified was equals with now() and because of this I put a small sleep there. This fixes the test.

jhford replied:
Relying on sleep() calls in tests concerns me a little.  Is there a reason we could do something like `new Date(now().getTime() - 1000)`?

elenasolomon replied:
The lastModified date is set by Azure and I can not control it.

jhford commented:
@jonasfj please do a final check.  I've also included a change on the [blob-storage-wip](https://github.com/taskcluster/fast-azure-storage/tree/blob-storage-wip) branch in #17 which drops support for 0.12 and bumps the version to 1.0.0

elenasolomon commented:
The code from authorizeWithRefreshSAS, authorizeWithSAS, authorizeWithSharedKey functions is 90% identical in all 3 services. Because of this, I moved them in a separate file 'authorization.js' and I changed a little bit the code to cover all the cases for Queue, Blob and Table. Can you please take a look on this and tell me if you agree with this solution? 
Thank you!

jhford commented:
Hey @jonasfj can you take a quick look at https://github.com/taskcluster/fast-azure-storage/pull/16#issuecomment-281916474 and merge this if you're OK with the solution that @elenasolomon has proposed?  Also, would you be so kind as to do the NPM release -- I'm not sure entirely what's involved in a fast-azure-storage release :)

jonasfj commented on lib/authorization.js:
```diff
@@ -0,0 +1,183 @@
+  };
+}
+
+exports.authorizeWithSharedKey = authorizeWithSharedKey;
```
nit, missing new line at end of file
